File: flutter_starter/test/domain/reminder_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:diario_crianca/domain/event.dart';
import 'package:diario_crianca/domain/reminder.dart';

void main() {
  final now = DateTime(2026, 7, 8, 12, 0);
  Event milkAgo(int minutes) => Event(
        id: 'm$minutes',
        type: EventType.milk,
        ts: now.subtract(Duration(minutes: minutes)),
        amountMl: 150,
      );
  Event evAgo(EventType type, int minutes) => Event(
        id: '${type.name}$minutes',
        type: type,
        ts: now.subtract(Duration(minutes: minutes)),
      );

  test('cenário 1: leite há 3h30, sem sono noturno → ATIVO', () {
    final r = milkReminderState([milkAgo(210)], now);
    expect(r.due, true);
    expect(r.active, true);
  });

  test('cenário 2: leite há 2h → não vencido', () {
    final r = milkReminderState([milkAgo(120)], now);
    expect(r.due, false);
    expect(r.active, false);
  });

  test('cenário 3: leite há 3h30 + soneca em aberto → ATIVO (soneca não suprime)', () {
    final r = milkReminderState([milkAgo(210), evAgo(EventType.napStart, 20)], now);
    expect(r.active, true);
  });

  test('cenário 4: leite há 3h30 + sono noturno em aberto → SUPRIMIDO', () {
    final r = milkReminderState([milkAgo(210), evAgo(EventType.nightStart, 40)], now);
    expect(r.due, true);
    expect(r.active, false);
  });

  test('cenário 5: sono noturno seguido de acordar → volta a ATIVAR', () {
    final r = milkReminderState([
      milkAgo(210),
      evAgo(EventType.nightStart, 300),
      evAgo(EventType.wakeMorning, 5),
    ], now);
    expect(r.active, true);
  });

  test('sem nenhum leite → inativo', () {
    final r = milkReminderState([evAgo(EventType.napStart, 10)], now);
    expect(r.active, false);
    expect(r.due, false);
  });

  test('night_start "esquecido" (>16h) não suprime para sempre', () {
    final r = milkReminderState([
      milkAgo(210),
      evAgo(EventType.nightStart, 20 * 60), // 20h atrás
    ], now);
    expect(r.active, true);
  });
}

File: test/domain/reminder.test.js
import { describe, it, expect } from "vitest";
import { milkReminderState, isNightSleeping } from "../../src/domain/reminder.js";

const now = new Date(2026, 6, 8, 12, 0).getTime();
const agoIso = (min) => new Date(now - min * 60000).toISOString();
const milkAgo = (min) => ({ id: `m${min}`, type: "milk", ts: agoIso(min), amountMl: 150 });
const evAgo = (type, min) => ({ id: `${type}${min}`, type, ts: agoIso(min) });

describe("milkReminderState", () => {
  it("cenário 1: leite há 3h30, sem sono noturno → ATIVO", () => {
    const r = milkReminderState([milkAgo(210)], now);
    expect(r.due).toBe(true);
    expect(r.active).toBe(true);
  });

  it("cenário 2: leite há 2h → não vencido", () => {
    const r = milkReminderState([milkAgo(120)], now);
    expect(r.due).toBe(false);
    expect(r.active).toBe(false);
  });

  it("cenário 3: leite há 3h30 + soneca em aberto → ATIVO (soneca não suprime)", () => {
    const r = milkReminderState([milkAgo(210), evAgo("nap_start", 20)], now);
    expect(r.active).toBe(true);
  });

  it("cenário 4: leite há 3h30 + sono noturno em aberto → SUPRIMIDO", () => {
    const r = milkReminderState([milkAgo(210), evAgo("night_start", 40)], now);
    expect(r.due).toBe(true);
    expect(r.active).toBe(false);
  });

  it("cenário 5: sono noturno seguido de acordar → volta a ATIVAR", () => {
    const r = milkReminderState(
      [milkAgo(210), evAgo("night_start", 300), evAgo("wake_morning", 5)],
      now,
    );
    expect(r.active).toBe(true);
  });

  it("sem nenhum leite → inativo", () => {
    const r = milkReminderState([evAgo("nap_start", 10)], now);
    expect(r.active).toBe(false);
    expect(r.due).toBe(false);
    expect(r.lastTs).toBeNull();
  });

  it('night_start "esquecido" (>16h) não suprime para sempre', () => {
    const r = milkReminderState([milkAgo(210), evAgo("night_start", 20 * 60)], now);
    expect(r.active).toBe(true);
  });
});

describe("isNightSleeping", () => {
  it("night_start em aberto (sem wake depois) = dormindo", () => {
    expect(isNightSleeping([evAgo("night_start", 60)], now)).toBe(true);
  });
  it("wake depois do night_start = acordado", () => {
    expect(isNightSleeping([evAgo("night_start", 300), evAgo("wake_morning", 5)], now)).toBe(false);
  });
});
